### PR TITLE
Ignore functional tensor wrapper when caching

### PR DESCRIPTION
**Pull request by jamesjwu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130252
* #129384
* #128583
* __->__ #128335

This PR makes it so that we don't try to serialize FunctionalTensorWrappers. FunctionalTensorWrappers don't pickle well because they have no underlying storage. This should be fixable at a later point, but I might not be the right author for implementing the serialization for it. If there's a way to avoid actually saving the FunctionalTensorWrappers themselves and just saving the ViewMetadata so we can replay it, that would also work.

To do this, we disable view_replay_input_mutations when using AOTAutogradCache, and then only keep the functional tensor in the ViewAndMutationMeta if we need it for view_replay_input_mutations (i.e. the cache is off).

### CODE REVIEW

File: torch/_functorch/config.py [hunk 1]
```diff
@@ -59,7 +62,10 @@
 # or default config to true and fix relevant bugs
 from torch._inductor.config import is_fbcode
 
-view_replay_for_aliased_outputs = not is_fbcode()
+# View replay is currently not compatible with AOTAutogradCache, since
+# FunctionalTensors are not serializable. We'll need to make them
+# serializable before enabling warm cache with this config turned on.
+view_replay_for_aliased_outputs = (not is_fbcode()) and (not enable_autograd_cache)
```
Comment (bdhirsh) [inline torch/_functorch/config.py:68]: fyi @ysiraichi @JackCaoG 

We're working on caching above the AOTAutograd layer as part of the effort to lower compile times, and right now this doesn't play well with Yukio's view-replay logic.

The problem is that we need to serialize any metadata that we need at runtime. Right now, the view-replay logic relies on grabbing the `std::vector<ViewMeta>` in `FunctionalTensorWrapper` at runtime in order to know what views to replay, but this whole thing is not serializable.

We're probably going to have to re-work the view-replay logic to save something else to the `ViewAndMutationMeta` that is more serializable (not the entire `FunctionalTensor`).

File: torch/_functorch/_aot_autograd/schemas.py [hunk 2]
```diff
@@ -176,7 +177,8 @@ class SubclassCreationMeta:
     # This is needed because we need the autograd metadata on the original subclass
     # (this is guaranteed to be a wrapper subclass that holds a fake tensor,
     #  so holding onto this at runtime shouldn't leak memory)
-    original_subclass: torch.Tensor
+    original_subclass: Optional[torch.Tensor]
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/schemas.py:180]: nit: probably worth a quick comment explaining why the type is Optional (actually required at compile time but not needed at runtime so we null it out)

**Review by bdhirsh — APPROVED** [review]

**Comment by jamesjwu** [comment]
Updated to recurse on nested tensor subclasses: `TestAOTAutograd.test_nested_subclasses_*` tests all pass, with or without the cache, so I think it should be working well

**Comment by jamesjwu** [comment]
@pytochbot merge

**Pull request closed without merging**